File: clients/java_graal/src/main/java/komus24/model/CityCell.java
package komus24.model;

import komus24.util.StreamUtil;

/**
 * TODO - Document
 */
public enum CityCell {
    /**
     * TODO - Document
     */
    ROAD(0),
    /**
     * TODO - Document
     */
    BUILDING(1),
    /**
     * TODO - Document
     */
    REFILL_STATION(2);

    public int tag;

    CityCell(int tag) {
        this.tag = tag;
    }

    /**
     * Read CityCell from input stream
     */
    public static CityCell readFrom(java.io.InputStream stream) throws java.io.IOException {
        switch (StreamUtil.readInt(stream)) {
        case 0:
            return ROAD;
        case 1:
            return BUILDING;
        case 2:
            return REFILL_STATION;
        default:
            throw new java.io.IOException("Unexpected tag value");
        }
    }
}

File: clients/scala/src/main/scala/komus24/model/CityCell.scala
package komus24.model

import komus24.util.StreamUtil

/**
 * City cell
 */
sealed abstract class CityCell (val tag: Int) {
    /**
     * Write CityCell to output stream
     */
    def writeTo(stream: java.io.OutputStream): scala.Unit = {
        StreamUtil.writeInt(stream, tag)
    }
}

object CityCell {
    /**
     * Road
     */
    case object ROAD extends CityCell(0)
    /**
     * Building
     */
    case object BUILDING extends CityCell(1)
    /**
     * Refill station
     */
    case object REFILL_STATION extends CityCell(2)

    /**
     * Read CityCell from input stream
     */
    def readFrom(stream: java.io.InputStream): CityCell =
        StreamUtil.readInt(stream) match {
            case ROAD.tag => ROAD
            case BUILDING.tag => BUILDING
            case REFILL_STATION.tag => REFILL_STATION
            case _ => throw new java.io.IOException("Unexpected tag value")
        }
}